highlight: add idea styling for code blocks and render scratchblocks client-side

arve0 opened the pull request:
ref issue #45 

Where should styling for the code block border and such go?

arve0 commented on src/assets/css/app.scss:
```diff
@@ -1 +1,2 @@
 @import "panels";
+@import "idea";
```
Is it possible to import from path? Tried `../../../node_modules/highlight.js/styles/idea.css`.

NorwegianKiwi commented:
Think so, panel imports from path.

arve0 commented:
Did not seem to work: 

```
The command "npm run test" exited with 0.
82.68s$ ./travis_build_prod.sh
Building production files while checking for errors...
clean-webpack-plugin: /home/travis/build/NorwegianKiwi/codeclub-viewer/dist has been removed.
ERROR in ./~/css-loader!./~/resolve-url-loader!./~/sass-loader?sourceMap!./~/bootstrap-loader/lib/bootstrap.styles.loader.js!./~/bootstrap-loader/no-op.js
ERROR in ./~/extract-text-webpack-plugin/loader.js?{"omit":1,"extract":true,"remove":true}!./~/style-loader!./~/css-loader!./~/resolve-url-loader!./~/sass-loader?sourceMap!./~/bootstrap-loader/lib/bootstrap.styles.loader.js!./~/bootstrap-loader/no-op.js
    ERROR in ./~/css-loader!./~/resolve-url-loader!./~/sass-loader?sourceMap!./~/bootstrap-loader/lib/bootstrap.styles.loader.js!./~/bootstrap-loader/no-op.js
Build FAILED!
```

arve0 commented:
Cryptic error message translated: wrong path in import statement. I had done a `npm install highlight` which put `markdown-it-loader`'s `highlight.js` inside `node_modules/markdown-it-loader/node_modules/highlight.js`.

NorwegianKiwi commented on src/assets/css/app.scss:
```diff
@@ -1 +1,2 @@
 @import "panels";
+@import "../../../node_modules/highlight.js/styles/idea";
```
Perhaps we should make a separate file called `_hljs.scss` og `_highlightjs.scss` and import that in app.scss, since I assume there will be more regarding highlighting.

Also, don't forget the newline at the end of the file ;)

arve0 commented:
As mentioned in #45, scratchblocks will only work through DOM-manipulation. How should we do that in a nice way?

arve0 commented:
> `ERROR in ./~/html-loader!./~/markdown-it-loader!./~/front-matter-loader?onlyBody!../oppgaver/src/scratch/3d_flakser/3d_flakser_1.md`

How can I get a more verbose error message?

kristandre commented:
Try `npm run build:prod`

arve0 commented:
Errors:

```
Module build failed: Error: Parse Error: <(distanse) < [1]>
Module build failed: Error: Parse Error: <(y) < [150]>
Module build failed: Error: Parse Error: <(x-posisjon) < [-235]>
Module build failed: Error: Parse Error: <(x-posisjon) > [240]>
Module build failed: Error: Parse Error: <(stoppet) = [0]>
Module build failed: Error: Parse Error: <(flaks) = [0]>
Module build failed: Error: Parse Error: < <tast [pil venstre v] trykket?> og ((blokkert venstre)=(0))>
Module build failed: Error: Parse Error: <(svar) = (riktig svar)>
Module build failed: Error: Parse Error: <(er generator) = [ja]>
Module build failed: Error: Parse Error: <(poeng) = [0]>
Module build failed: Error: Parse Error: <(bakgrunnsnavn) = [meny]>
Module build failed: Error: Parse Error: <ikke <berører [Helikopter v]?>>
Module build failed: Error: Parse Error: <(avstand til [musepeker v]) > [10]>
Module build failed: Error: Parse Error: <(y-posisjon) > [190]>
Module build failed: Error: Parse Error: <(tilfeldig tall fra (1) til (25)) = [1]>
Module build failed: Error: Parse Error: <(steg) > [1]>
Module build failed: Error: Parse Error: <ikke <berører [kant v]?>>
Module build failed: Error: Parse Error: <(opptak) = [1]>
Module build failed: Error: Parse Error: <(Tid) = [0]>
Module build failed: Error: Parse Error: <(kappløp) = [1]>
Module build failed: Error: Parse Error: <<tast [w v] trykket> og <(y-posisjon) < [150]>>
Module build failed: Error: Parse Error: <(Poeng) = (Mål)>
Module build failed: Error: Parse Error: <((teller) - (lengde)) > (min id)>
Module build failed: Error: Parse Error: <ikke <tast [mellomrom v] trykket?>>
Module build failed: Error: Parse Error: <<berører [Katt v] ?> og <(Flytt) > [20]>>
Module build failed: Error: Parse Error: <(tid) = [0]>
Module build failed: Error: Parse Error: <(x-posisjon) > [160]>
Module build failed: Error: Parse Error: <<<(mus x) > [-230]> og <(mus x) < [230]>> og <<(mus y) > [-120]> og <(mus y) < [170]>>>
```

This is not valid HTML, so it is correct to fail/complain. A node SVG-renderer would solve this. Any other ideas?

NorwegianKiwi commented:
I would like to take a look. After all your comments, and the suggested changes in this PR, I have a much better idea of what is going on.

arve0 commented:
I've made some progress. One can measure all characters in browser up front, then store character lengths in a map.

NorwegianKiwi commented:
PS: If we are going to use Promises, we need to include them. I suggest bluebird, but open to suggestions.

arve0 commented:
> PS: If we are going to use Promises, we need to include them. I suggest bluebird, but open to suggestions.

We do not. This code is one time run only, which should create a map like:

```
{
  a: 6,
  b: 7,
  ...
}
```

Where 6 and 7 are widths of char a and b.

arve0 commented:
Getting there...

![skjermbilde 2016-07-08 kl 01 46 13](https://cloud.githubusercontent.com/assets/4810521/16673375/d5209edc-44ad-11e6-9c57-075f9923f63a.png)

![skjermbilde 2016-07-08 kl 01 47 12](https://cloud.githubusercontent.com/assets/4810521/16673385/ec0cab0e-44ad-11e6-85ac-7c0d52b5cc50.png)

Some trouble with font, some strings are reported to be shorter then they are. [The code](https://github.com/arve0/scratchblocks/tree/node) is a bit hacky right now, but should be quite easy to clean up.

arve0 commented:
Got it working!

One obvious caveat: measuring the text size server-side will depend on fonts installed. We must assure that fonts are available in the clients CSS, to assure client and server uses the same font. Or we should **force rerender** on client. 

Current font-family is `Lucida Grande,Verdana,Arial,DejaVu Sans,sans-serif` for key words and `Helevetica,Arial,DejaVu Sans,sans-serif` for comments.

NorwegianKiwi commented:
That is great news! (Crazy man sacrificing his night for the project :-) 👍 )
I am looking forward to taking a look at the code as soon as I get a chance!

arve0 commented:
This is done, but text **will** render badly when server and client does not have the same font. System fonts is used:

OS X: Lucida Grande
Windows: Lucida Sans Unicode
Linux/Android: Some variant of sans, Verdana, Arial, sans-serif

For more information, see [fontfamily.io](http://fontfamily.io/Lucida_Grande,Lucida_Sans_Unicode,Verdana,Arial,sans-serif;)

arve0 commented:
Note, https://github.com/NorwegianKiwi/codeclub-viewer/pull/70/commits/51b1f47c4336daf71ff373770f2d86623180a24f would be solution which works everywhere.

NorwegianKiwi commented:
Looks like there is something incompatible when doing npm install. Have you done a "nvm use" locally to make sure node is at version 6.2.2?

arve0 commented:
I was on `v6.2.1`, do we have such a locked dependency on bleeding node version?

Though, nothing fails here on `npm install`: https://asciinema.org/a/7sid0an9qf11str21kwtthgir

arve0 commented:
Removed the master merge, as github is not able to detect the solved merge conflict. Did a rebase instead.

Added dependencies on c++ compiler and libraries for [cairo](https://github.com/Automattic/node-canvas#installation).

arve0 commented:
I'm giving up this. I think we should go with 51b1f47.

If you would like to take this further, webpack must be configured to bundle the `scratchblocks` module in `node_modules/scratchblocks/browser/scratchblocks.js`, which is a browser build. This is because `scratchblocks` uses some [tricks](https://github.com/arve0/scratchblocks/blob/master/src/scratchblocks.js#L15-L29) to avoid bundling modules which should be available in the browser and still functioning as it should on node.

NorwegianKiwi commented:
I think perhaps you are right that it might be better to not prerender the blocks on the server, but rather find the best way of including a script in the Lesson component. I tried adding the changes in commit 51b1f47 in my own branch, but it seems that was not enough, since it could not find `<pre class="blocks">` or `<code class="b">`, but I probably haven't configured the hightlighting preprosessing correctly (yet).

Anyway, perhaps we could talk about this when you have some time, so that I can get an even better understanding, @arve0 ?

NorwegianKiwi commented:
Closing this PR, continuing the work on PR #140.

arve0 commented:
Not thinking about this, seemed to work. I got an idea 😉 

I feel this is a better approach then `scratchblock.renderMatching` which alters the DOM. `componentWillMount` will run both server- and client-side. This will replace HTML content before it is is mounted to the DOM.

arve0 commented:
This will only work with the lates @arve0/scratchblocks#master. The latest scratchblocks in arve0 contains a new browser build also has a webpack entry-point, `require('scratchblocks/browser/index.js')`. But webpack seem to only remove unreachable branches of code on `webpack -p`(?), so I used the browserified build. Webpack complains about this in the console..

HTML-lint will still fail.

NorwegianKiwi commented:
I am having problems with `npm install`. I get the following errors:
```
Lances-MacBook-Pro:codeclub-viewer lance$ npm install
npm WARN prefer global colorguard@1.2.0 should be installed with -g
npm WARN prefer global node-gyp@3.4.0 should be installed with -g

> fsevents@1.0.14 install /Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/fsevents
> node-pre-gyp install --fallback-to-build

[fsevents] Success: "/Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/fsevents/lib/binding/Release/node-v48-darwin-x64/fse.node" is installed via remote

> canvas@1.4.0 install /Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/canvas
> node-gyp rebuild

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
  SOLINK_MODULE(target) Release/canvas-postbuild.node
  CXX(target) Release/obj.target/canvas/src/Canvas.o
In file included from ../src/Canvas.cc:20:
../src/JPEGStream.h:10:10: fatal error: 'jpeglib.h' file not found
#include <jpeglib.h>
         ^
1 error generated.
make: *** [Release/obj.target/canvas/src/Canvas.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/lance/.nvm/versions/node/v6.2.2/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)
gyp ERR! System Darwin 15.6.0
gyp ERR! command "/Users/lance/.nvm/versions/node/v6.2.2/bin/node" "/Users/lance/.nvm/versions/node/v6.2.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/canvas
gyp ERR! node -v v6.2.2
gyp ERR! node-gyp -v v3.3.1
gyp ERR! not ok
npm WARN install:canvas@1.4.0 canvas@1.4.0 install: `node-gyp rebuild`
npm WARN install:canvas@1.4.0 Exit status 1

> node-sass@3.7.0 install /Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/node-sass
> node scripts/install.js

Binary downloaded and installed at /Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/node-sass/vendor/darwin-x64-48/binding.node

> node-sass@3.7.0 postinstall /Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/node-sass
> node scripts/build.js

"/Users/lance/Projects/kodeklubben/codeclub-viewer/node_modules/node-sass/vendor/darwin-x64-48/binding.node" exists.
 testing binary.
Binary is fine; exiting.
codeclub-viewer@1.0.0 /Users/lance/Projects/kodeklubben/codeclub-viewer
+-- autoprefixer@6.3.6
| +-- browserslist@1.3.5
| +-- caniuse-db@1.0.30000507
...
...
```

I am suspecting that it has something to do with scratchblocks using canvas. At least it says on the webpage (https://github.com/Automattic/node-canvas) that I must install some packages (did so using brew), and I did so, but it still gives me these errors. Am I missing something?

NorwegianKiwi commented:
Also, would be interesting if master was merged in to see what failed here.

NorwegianKiwi commented:
Hm, looks like I just forgot to do `xcode-select --install`. After doing that, and doing `npm install`again, it works.

NorwegianKiwi commented:
By the way, if we need to install cairo to have this work (as described in https://github.com/Automattic/node-canvas), this should probably be mentioned in the README-file.

arve0 commented:
node-canvas should be optional:

```
  "optionalDependencies": {
    "canvas": "1.4.0",
    "node-fetch": "^1.5.3",
    "xmldom": "^0.1.22"
  }
```
https://github.com/arve0/scratchblocks/blob/master/package.json

arve0 commented:
Is the build doing some static analysis? I get this error:

```
ERROR in ReferenceError: document is not defined
    at newCanvas (staticbundle:31283:13)
    at measureText (staticbundle:32766:61)
    at staticbundle:32811:22
    at Array.forEach (native)
    at measureAllLabels (staticbundle:32802:14)
```

but the [`newCanvas` function](https://github.com/arve0/scratchblocks/blob/master/browser/scratchblocks.js#L4340) should never run serverside. `scratchblocks` are only called here: https://github.com/arve0/codeclub-viewer/blob/highlight/src/components/Lesson/Lesson.js#L28

arve0 commented on README.md:
````diff
@@ -11,6 +11,15 @@ npm start
 ```
 Then open http://localhost:8080
 
+**Note:** You will get an node-gyp error
````
**a** node-gype error

arve0 commented:
Haha, logical error:

```
if (!process) {
  return;
}
```

arve0 commented on src/components/Lesson/Lesson.js:
```diff
@@ -15,6 +18,16 @@ const Lesson = React.createClass({
       __html: this.props.lesson.content
     };
   },
+  componentWillMount(){
```
This function minifies to:

```
            componentWillMount: function() {
                if (!t) {
                    var e = this.props.lesson.content;
                    a(e) && (this.props.lesson.content = o(e))
                }
            },
```

Setting a breakpoint at `if (!t)` yields the value of `t` from the closure:

![skjermbilde 2016-07-31 kl 14 56 50](https://cloud.githubusercontent.com/assets/4810521/17276643/3df274d8-572f-11e6-9764-2a2a5fe7755d.png)

Any ideas? Some webpack wizardry?

arve0 replied:
https://webpack.github.io/docs/api-in-modules.html#process

NorwegianKiwi replied:
I am not sure if I follow now. Are you confused because `process` is an object with a lot of stuff? I though process was something nodejs used?

Anyway, why not just check if `document` is defined to see if we are running a server process?
```
if (typeof document !== 'undefined') {
  // Client render code goes here...
}
```

Or, in `webpack.static.config.babel.js` just use `webpack.DefinePlugin()` to define a variable to say that we are running the server version?
https://webpack.github.io/docs/list-of-plugins.html#defineplugin

NorwegianKiwi commented:
I assume you are aware, @arve0, but putting it here just to be clear. It seems that the code now builds (also static), but I see no scratchblocks. I haven't checked why. Also, I get the following warning during compile:
```
WARNING in ./~/scratchblocks/browser/scratchblocks.js
Critical dependencies:
1:486-493 This seems to be a pre-built javascript file. Though this is possible, it's not recommended. Try to require the original source to get better results.
 @ ./~/scratchblocks/browser/scratchblocks.js 1:486-493
```

Are you stuck, do you need specific answers? Or are you in the middle of working on this, and planning to do more?

arve0 commented:
Yes, I'm aware. process object is available client-side, passed to the
closure. I'll try your suggestion.

Unsure what to do with the warning, as webpack does not remove unreachable
branches when in development mode, only the production build removes
unreachable branches of code. This breaks even more than the warning you
are seeing.

arve0 commented:
This function does not work on production build (but does work on dev server):

```
function containsScratchCode(html){
  return (html.indexOf('<pre class="blocks">') !== -1 ||
          html.indexOf('<code class="b">') !== -1);
}
```

because

```
<pre class="blocks">
```

is minified to

```
<pre class=blocks>
```

What would be the way to fix that?

arve0 commented:
Rewriting as:

```
function containsScratchCode(html){
  let blocks = [
    '<pre class="blocks">',
    '<pre class=blocks>',  // minified in production build
    '<code class="b">',
    '<code class=b>'
  ];
  for (let i = 0; i < blocks.length; i += 1) {
    if (html.indexOf(block) !== -1) {
      return true;
    } 
  }
  return false;
}
```

arve0 commented:
Mama mia. I think this is ready to go 🎉

NorwegianKiwi commented:
Fantastic!

NorwegianKiwi commented:
If you add the following line to `webpack.base.config.babel.js` in the "module"-section right before "loaders" (line 97) we also get rid of the warning that we are using precomiled javascript:
```
noParse: /node_modules\/scratchblocks\/browser\/scratchblocks.js/,
```

Perhaps Lesson.js could even include the minified version? (would need to change the regexp above of course).

NorwegianKiwi commented:
By the way (just curious), but what is the point of these optional dependencies? I mean, why would you install something if there is a workaround in case it fails? Why not always use the workaround? If not, you risk getting different results depending on whether you successfully installed the optional stuff...

NorwegianKiwi commented:
If you agree that we should add the "noParse" line, and you do, I will merge this!

arve0 commented:
Agree on noParse.

arve0 commented:
> what is the point of these optional dependencies?

They are optional because one may use scratchblocks in node. Then they are needed. We only use scratchblocks in client where we have a DOM with canvas available. Of course, this could be broken up in two modules, like scratchblocks-node and scratchblocks-client.

arve0 commented:
On iPad now, so feel free to fix the warning yourself.

NorwegianKiwi commented:
PR #155 is a squashed version of this, and has been merged. Thanks for not giving up, @arve0 !
